File: implementation/20170606/02/main.cpp
#include<iostream>

using namespace std;

int main() {
  ios::sync_with_stdio(false);
  long long x, y, l, r, count, max = 0;
  cin >> x >> y >> l >> r;
  bool unlucky[r+2];
  unlucky[r+1] = true;
  for (long long current = 1; current < r; current *= x) {
    if (current < l-1) continue;
    for (long long second = 1; (current + second) <= r; second *= y) {
      unlucky[current + second] = true;
    }
  }
  for (long long current = 1; current < r; current *= y) {
    if (current < l-1) continue;
    for (long long second = 1; (current + second) <= r; second *= x) {
      unlucky[current + second] = true;
    }
  }
  count = 0;
  for(long long i = l; i <= r+1; i++) {
    if (unlucky[i]) {
      max = max > count ? max : count;
      count = -1;
    }
    count++;
  }
  cout << max << endl;
  return 0;
}

File: implementation/20170606/02/Main.java
import java.util.Scanner;
import java.util.Set;
import java.util.TreeSet;

public class Main {

  public static void main(String[] args) {
    Scanner in = new Scanner(System.in);
    long last, max = 0;
    long x = in.nextLong();
    long y = in.nextLong();
    long l = in.nextLong();
    long r = in.nextLong();
    Set<Long> unlucky = new TreeSet<Long>();
    unlucky.add(r+1);
    for (long current = 1; current > 0 && current < r; current *= x) {
      if ( current != 1 && current%x != 0) break;
      for (long second = 1; second > 0 && (current + second) <= r; second *= y) {
        if ( second != 1 && second%y != 0) break;
        if ( (current + second) < l) continue;
        unlucky.add(current+second);
      }
    }
    last = l-1;
    for (long current : unlucky) {
      if ( last + 1 != current) {
        max = Math.max(max, current -1 - last);
      }
      last = current;
    }
    System.out.println(max);
    in.close();
  }
}
